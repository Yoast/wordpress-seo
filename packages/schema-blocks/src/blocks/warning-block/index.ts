import { WarningBlockAttributes } from "./warningBlockAttributes";

export { WarningBlockAttributes };
