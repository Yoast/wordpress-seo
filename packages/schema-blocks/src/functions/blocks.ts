import { select } from "@wordpress/data";

/**
 * Gets attributes of inner blocks.
 *
 * @param clientId The client ID of the parent block.
 * @param blocks   A mapping of block name to attribute key.
 *
 * @returns An array contain block names and values.
 */
export function getInnerBlocksAttributes( clientId: string, blocks: Record<string, string> ): { name: string; value: unknown }[] {
	let innerBlocks = select( "core/block-editor" ).getBlock( clientId ).innerBlocks;
	innerBlocks     = innerBlocks.filter( block => block.name in blocks );

	const values = [];

	for ( const block of innerBlocks ) {
		const key   = blocks[ block.name ];
		values.push( { name: block.name, value: block.attributes[ key ] } );
	}

	return values;
}
