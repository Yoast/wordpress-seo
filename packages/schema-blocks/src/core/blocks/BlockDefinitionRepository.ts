import BlockDefinition from "./BlockDefinition";

// Internal store of all known BlockDefinitions.
const registeredBlockDefinitions: Record<string, BlockDefinition> = {};

/**
 * Finds a registered BlockDefinition.
 *
 * @param name The Block to search for.
 *
 * @returns {BlockDefinition} The found block.
 */
export function getBlockDefinition( name: string ): BlockDefinition {
	return registeredBlockDefinitions[ name ];
}

/**
 * Registers a BlockDefinition.
 *
 * @param name       The name of the Definition to register.
 * @param definition The Definition to register.
 */
export function registerBlockDefinition( name: string, definition: BlockDefinition ): void {
	registeredBlockDefinitions[ name ] = definition;
}
