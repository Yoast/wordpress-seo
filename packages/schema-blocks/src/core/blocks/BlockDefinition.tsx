import { ReactElement } from "react";
import { createElement, Fragment } from "@wordpress/element";
import {
	registerBlockType,
	BlockConfiguration,
	BlockEditProps,
	BlockSaveProps,
} from "@wordpress/blocks";
import { InspectorControls, BlockIcon } from "@wordpress/block-editor";
import { select } from "@wordpress/data";
import { PanelBody } from "@wordpress/components";

import BlockInstruction from "./BlockInstruction";
import Definition from "../Definition";
import BlockRootLeaf from "../../leaves/blocks/BlockRootLeaf";
import parse from "../../functions/blocks/parse";
import { registerBlockDefinition } from "./BlockDefinitionRepository";
import logger from "../../functions/logger";
import { openGeneralSidebar } from "../../functions/gutenberg/sidebar";

export interface RenderEditProps extends BlockEditProps<Record<string, unknown>> {
	clientId: string;
	name?: string;
}

export interface RenderSaveProps extends BlockSaveProps<Record<string, unknown>> {
	clientId?: string;
	className?: string;
}

export type MutableBlockConfiguration = {
	-readonly [K in keyof BlockConfiguration]: BlockConfiguration[K]
}

/**
 * BlockDefinition class.
 */
export default class BlockDefinition extends Definition {
	public static separatorCharacters = [ "b", "c", "d", "f", "g", "h", "k", "m", "z" ];
	public static parser = parse;

	public instructions: Record<string, BlockInstruction>;
	public tree: BlockRootLeaf;

	/**
	 * Renders editing the block.
	 *
	 * @param props The props.
	 *
	 * @returns The rendered block.
	 */
	edit( props: RenderEditProps ): ReactElement {
		// Force the sidebar open.
		if ( select( "core/block-editor" ).isBlockSelected( props.clientId ) ) {
			openGeneralSidebar( "edit-post/block", true );
		}

		const sidebarElements = this.sidebarElements( props );

		// Take the children directly to avoid creating too many Fragments.
		const elements = this.tree.children.map( ( leaf, i ) => leaf.edit( props, i ) ).filter( e => e !== null );

		if ( sidebarElements.length > 0 ) {
			const sidebarContainer =
				<InspectorControls key="sidebar-inspector-controls">
					<PanelBody>
						{ sidebarElements }
					</PanelBody>
				</InspectorControls>;

			elements.unshift( sidebarContainer );
		}

		if ( elements.length === 1 ) {
			return elements[ 0 ] as ReactElement;
		}

		return createElement( Fragment, { key: props.clientId }, elements.map( ( element, i ) => createElement( Fragment, { key: i }, element ) ) );
	}

	/**
	 * Renders the persisted block.
	 *
	 * @param props The props.
	 *
	 * @returns The rendered block.
	 */
	save( props: RenderSaveProps ): ReactElement {
		return this.tree.save( props );
	}

	/**
	 * Registers the block with Gutenberg.
	 */
	register(): void {
		const configuration = this.configuration() as MutableBlockConfiguration;
		const name = configuration.name as string;
		delete configuration.name;

		configuration.edit = props => this.edit( props );
		configuration.save = props => this.save( props );

		logger.debug( "registering block " + name );

		if ( configuration.icon && typeof configuration.icon === "string" && configuration.icon.startsWith( "<svg" ) ) {
			configuration.icon = this.createBlockIcon( configuration );
		}

		// Register the block to WordPress.
		registerBlockType( name, configuration );
		// Register the block with our own code.
		registerBlockDefinition( name, this );
	}

	/**
	 * Creates the sidebar elements.
	 *
	 * @param props The properties of the block to create a sidebar for.
	 *
	 * @returns The sidebar element to render.
	 */
	sidebarElements( props: RenderEditProps ): ReactElement[] {
		return Object.values( this.instructions )
			.map( ( instruction, index ) => <Fragment key={ instruction.id }>
				{ instruction.sidebar( props, index ) }
			</Fragment> )
			.filter( e => e !== null );
	}

	/**
	 * Creates a block icon.
	 *
	 * @param configuration The block configuration.
	 *
	 * @returns The sidebar element to render.
	 */
	private createBlockIcon( configuration: MutableBlockConfiguration ): ReactElement {
		const icon = <span
			className="yoast-schema-blocks-icon"
			dangerouslySetInnerHTML={ { __html: configuration.icon as string } }
		/>;

		return <BlockIcon icon={ icon } />;
	}
}
