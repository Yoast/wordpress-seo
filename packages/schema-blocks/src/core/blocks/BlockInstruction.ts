import BlockLeaf from "./BlockLeaf";
import { RenderEditProps, RenderSaveProps } from "./BlockDefinition";
import { ReactElement } from "react";
import { BlockConfiguration, BlockInstance } from "@wordpress/blocks";
import { BlockValidation, BlockValidationResult, BlockPresence } from "../validation";
import Instruction from "../Instruction";

/**
 * BlockInstruction class.
 */
export default abstract class BlockInstruction extends Instruction {
	/* eslint-disable @typescript-eslint/no-unused-vars */
	/**
	 * Renders saving the element.
	 *
	 * @param props The props.
	 * @param leaf  The leaf being rendered.
	 * @param i     The number the rendered element is of its parent.
	 *
	 * @returns {ReactElement | string} The element to render.
	 */
	save( props: RenderSaveProps, leaf: BlockLeaf, i: number ): ReactElement | string {
		return null;
	}
	/* eslint-enable @typescript-eslint/no-unused-vars */

	/* eslint-disable @typescript-eslint/no-unused-vars */
	/**
	 * Renders editing the element.
	 *
	 * @param props The props.
	 * @param leaf  The leaf being rendered.
	 * @param i     The number the rendered element is of its parent.
	 *
	 * @returns {ReactElement | string} The element to render.
	 */
	edit( props: RenderEditProps, leaf: BlockLeaf, i: number ): ReactElement | string {
		return null;
	}
	/* eslint-enable @typescript-eslint/no-unused-vars */

	/* eslint-disable @typescript-eslint/no-unused-vars */
	/**
	 * Renders the sidebar.
	 *
	 * @param props The props.
	 * @param i     The number the rendered element is of its parent.
	 *
	 * @returns {ReactElement | string} The sidebar element to render.
	 */
	sidebar( props: RenderEditProps, i: number ): ReactElement {
		return null;
	}
	/* eslint-enable @typescript-eslint/no-unused-vars */

	/**
	 * Returns the configuration of this instruction.
	 *
	 * @returns The block configuration.
	 */
	configuration(): Partial<BlockConfiguration> {
		return {
			attributes: {
				[ this.options.name ]: {
					required: this.options.required === true,
				},
			},
		};
	}

	/**
	 * Checks if the instruction block is valid.
	 *
	 * @param blockInstance The attributes from the block.
	 *
	 * @returns {BlockValidationResult} The validation result.
	 */
	validate( blockInstance: BlockInstance ): BlockValidationResult {
		return new BlockValidationResult( blockInstance.clientId, this.constructor.name, BlockValidation.Unknown, BlockPresence.Unknown );
	}
}
