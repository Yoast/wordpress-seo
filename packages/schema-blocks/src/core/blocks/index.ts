import BlockDefinition from "./BlockDefinition";
import BlockInstruction from "./BlockInstruction";
import { BlockInstructionClass } from "./BlockInstructionClass";
import BlockLeaf from "./BlockLeaf";
import * as BlockDefinitionRepository from "./BlockDefinitionRepository";

export {
	BlockDefinition,
	BlockInstruction,
	BlockInstructionClass,
	BlockLeaf,
	BlockDefinitionRepository,
};
