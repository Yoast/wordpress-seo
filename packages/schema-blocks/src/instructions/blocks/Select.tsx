import { ReactElement } from "react";
import { createElement, useCallback } from "@wordpress/element";
import { BlockInstance } from "@wordpress/blocks";
import { SelectControl } from "@wordpress/components";
import { BlockInstruction } from "../../core/blocks/";
import { RenderEditProps, RenderSaveProps } from "../../core/blocks/BlockDefinition";
import { BlockValidationResult } from "../../core/validation";
import { defaultValidate } from "../../functions/validators/defaultValidate";

/**
 * A name of a standard HTML element.
 * E.g. "span" or "div".
 */
type IntrinsicElement = keyof JSX.IntrinsicElements;

/**
 * Select (a drop-down box) instruction.
 */
export default class Select extends BlockInstruction {
	public options: {
		/**
		 * The attribute name the value selected in the select control should be saved as.
		 */
		name: string;
		/**
		 * The available options within the select control.
		 */
		options: { label: string; value: string }[];
		/**
		 * The label that should be shown before this select control.
		 */
		label: string;
		/**
		 * If it is required that a value is selected.
		 */
		required?: boolean;
		/**
		 * Whether to visually hide the label.
		 */
		hideLabelFromVision: boolean;
		/**
		 * An optional extra class name or class names.
		 */
		className?: string;
		/**
		 * The default selected value.
		 */
		defaultValue?: string;
		/**
		 * The tagname to render the output as.
		 */
		tag?: IntrinsicElement;
	};

	/**
	 * Renders saving the element.
	 *
	 * @param props The props.
	 *
	 * @returns {JSX.Element} The element to render.
	 */
	save( props: RenderSaveProps ): ReactElement | string {
		const { label, name, hideLabelFromVision } = this.options;

		const value = props.attributes[ name ] as string;
		let TagName: IntrinsicElement = "span";
		if ( this.options.tag ) {
			TagName = this.options.tag;
		}

		return <TagName data-id={ name } data-value={ value }>
			{ ! hideLabelFromVision && <strong>{ label }:</strong> }
			{ this.label( value ) }
		</TagName>;
	}

	/**
	 * Checks if the instruction block is valid.
	 *
	 * @param blockInstance The attributes from the block.
	 *
	 * @returns {BlockValidationResult} The validation result.
	 */
	validate( blockInstance: BlockInstance ): BlockValidationResult {
		return defaultValidate( blockInstance, this );
	}

	/**
	 * Returns the label of the selected option.
	 *
	 * @param value The render props.
	 *
	 * @returns The label of the selected option.
	 */
	protected label( value: string ): string {
		const foundOption = this.options.options.find( option => option.value === value );
		if ( foundOption ) {
			return foundOption.label;
		}
		return null;
	}

	/**
	 * Renders editing the element.
	 *
	 * @param props The props.
	 *
	 * @returns {JSX.Element} The element to render.
	 */
	edit( props: RenderEditProps ): ReactElement | string {
		const { label, options, hideLabelFromVision, className, defaultValue } = this.options;

		const value = props.attributes[ this.options.name ] as string;

		if ( ! value ) {
			props.setAttributes( { [ this.options.name ]: defaultValue || options[ 0 ].value } );
		}

		/**
		 * Function that is called whenever a new value is selected in the select element.
		 *
		 * @param event The change event.
		 */
		const onChange = useCallback(
			newValue => {
				props.setAttributes( { [ this.options.name ]: newValue } );
			},
			[ props ],
		);

		return <SelectControl
			className={ [ className, "yoast-schema-select" ].join( " " ) }
			label={ label }
			value={ value }
			defaultValue={ defaultValue }
			onChange={ onChange }
			options={ options }
			hideLabelFromVision={ hideLabelFromVision }
		/>;
	}
}

BlockInstruction.register( "select", Select );
