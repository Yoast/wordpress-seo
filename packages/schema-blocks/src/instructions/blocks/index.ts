import "./Block";
import "./ClassName";
import "./CurrencySelect";
import "./Date";
import "./Heading";
import "./InheritSidebar";
import "./InnerBlocks";
import "./RichText";
import "./Select";
import "./SidebarCheckbox";
import "./SidebarDuration";
import "./SidebarInput";
import "./SidebarSelect";
import "./TextInput";
import "./Title";
import "./Variation";
import "./VariationPicker";
import "./Duration";
