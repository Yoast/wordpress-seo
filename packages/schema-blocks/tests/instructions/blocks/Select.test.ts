import * as renderer from "react-test-renderer";

import { BlockConfiguration } from "@wordpress/blocks";

import Select from "../../../src/instructions/blocks/Select";
import { RenderSaveProps } from "../../../src/core/blocks/BlockDefinition";
import { RenderEditProps } from "../../../src/core/blocks/BlockDefinition";

describe( "The Select instruction", () => {
	const options: Select["options"] = {
		name: "cuisine",
		label: "Cuisine",
		options: [
			{ label: "Korean", value: "korean" },
			{ label: "Tanzanian", value: "tanzanian" },
			{ label: "Australian", value: "australian" },
		],
		hideLabelFromVision: false,
	};

	describe( "the save method", () => {
		it( "renders the correct React tree.", () => {
			const props: RenderSaveProps = {
				attributes: {
					cuisine: "korean",
				},
			};

			const selectInstruction = new Select( 123, options );

			const tree = renderer
				.create( selectInstruction.save( props ) as any )
				.toJSON();

			expect( tree ).toMatchSnapshot();
		} );
	} );

	describe( "the configuration method", () => {
		it( "returns the correct configuration.", () => {
			const newOptions = Object.assign( {}, options, { required: true } );

			const expectedConfiguration: Partial<BlockConfiguration> = {
				attributes: {
					cuisine: {
						required: true,
					},
				},
			};

			const selectInstruction = new Select( 123, newOptions );

			expect( selectInstruction.configuration() ).toEqual( expectedConfiguration );
		} );
	} );

	/**
	 * Skipped because React hooks do not work in the tests yet.
	 */
	describe.skip( "The edit method", () => {
		it( "renders the correct React tree", () => {
			const props: RenderEditProps = {
				clientId: "abcd-1234",
				className: "",
				attributes: {
					cuisine: "korean",
				},
				isSelected: true,
				setAttributes: jest.fn(),
			};

			const selectInstruction = new Select( 123, options );

			const tree = renderer
				.create( selectInstruction.edit( props ) as any )
				.toJSON();

			expect( tree ).toMatchSnapshot();
		} );
	} );
} );
