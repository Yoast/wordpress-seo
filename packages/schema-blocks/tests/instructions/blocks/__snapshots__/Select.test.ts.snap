// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Select instruction the save method renders the correct React tree. 1`] = `
<span
  data-id="cuisine"
  data-value="korean"
>
  <strong>
    Cuisine
    :
  </strong>
  Korean
</span>
`;
